### PR TITLE
Jvenstad/recursive application api

### PR DESCRIPTION
**Pull request by jonmv**
@bratseth please review. 

You can now provide "recursive=true" at the paths:

1. `/application/v4/`
2. `/application/v4/tenant/someone/`
3. `/application/v4/tenant/someone/application/someapp`

At the application level, this lists full information about all deployments of the application, with deployment instance (argh), region and environment.

At the tenant level, this lists all applications for the tenant, as above, with application name. 

At the root level, this lists all tenants, as above, with tenant name. 

Obviously, examples are found in the test responses, like `recursive-root.json`.

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 1]
```diff
@@ -222,10 +221,22 @@ private HttpResponse handleOPTIONS() {
...
+    private HttpResponse recursiveRoot(HttpRequest request) {
+        Slime slime = new Slime();
+        Cursor tenantArray = slime.setArray();
+        for (Tenant tenant : controller.tenants().asList()) {
+            Cursor tenantObject = tenantArray.addObject();
+            tenantObject.setString("tenant", tenant.getId().id());
+            toSlime(tenantObject, tenant, request, true);
+        }
+        return new SlimeJsonResponse(slime);
+    }
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: New entry point.

**Comment by jonmv** [comment]
The diff is misleadingly large, because I moved stuff around a little. I'll mark the real changes, and try to separate these kinds of changes next time.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 2]
```diff
@@ -222,10 +221,22 @@ private HttpResponse handleOPTIONS() {
...
+
     private HttpResponse root(HttpRequest request) {
-        return new ResourceResponse(request, "user", "tenant", "tenant-pipeline", "athensDomain",
-                                    "property", "cookiefreshness");
+        return request.getBooleanProperty("recursive")
+                ? recursiveRoot(request)
+                : new ResourceResponse(request, "user", "tenant", "tenant-pipeline", "athensDomain", "property", "cookiefreshness");
     }
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: Change behaviour if `recursive=true`.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 3]
```diff
@@ -385,33 +388,36 @@ else if (((Change.ApplicationChange)application.deploying().get()).revision().is
...
                 setRotationStatus(deployment, rotationHealthStatus, deploymentObject);
-            deploymentObject.setString("url", withPath(path.asString() +
-                                                       "/environment/" + deployment.zone().environment().value() +
-                                                       "/region/" + deployment.zone().region().value() +
-                                                       "/instance/" + application.id().instance().value(),
-                                                       request.getUri()).toString());
+
+            if (request.getBooleanProperty("recursive")) // List full deployment information when recursive.
+                toSlime(deploymentObject, new DeploymentId(application.id(), deployment.zone()), deployment, request);
+            else
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: Change behaviour if `recursive=true`.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 4]
```diff
@@ -989,29 +1012,14 @@ private static String path(Object... elements) {
...
-    }
-
     private void toSlime(Application application, Cursor object, HttpRequest request) {
         object.setString("application", application.id().application().value());
         object.setString("instance", application.id().instance().value());
-        object.setString("url", withPath("/application/v4/tenant/" + application.id().tenant().value() +
-                                         "/application/" + application.id().application().value(), request.getUri()).toString());
+        if (request.getBooleanProperty("recursive"))
+            toSlime(object, application, request);
+        else
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: Change behaviour if `recursive=true`.

**Comment by bratseth** [comment]
LGTM but could you use @include instead in the .json expected responses?
Two advantages of this: We validate that it is the same, and we don't have to change even more places when we add fields.

**Comment by bratseth** [comment]
For example, the nodes/v2 API is recursive and the expected response json file for recursive nodes reads

```
{
  "nodes": [
    @include(node7.json),
    @include(parent1.json),
    @include(node10.json),
    @include(node4.json),
    @include(node6.json),
    @include(node3.json),
    @include(node2.json),
    @include(node1.json),
    @include(node55.json),
    @include(node5.json)
  ]
}
```

**Comment by jonmv** [comment]
I'll most certainly do that!

**Comment by jonmv** [comment]
I suppose it's also possible to just put the content of the tenant object in a file, and use 
`{ @include(someTenant.json) }` for the tenant response, and
`{ "tenant": someTenant, @include(someTenant.json) }` where it is in a list.

**Comment by jonmv** [comment]
Ok, as discussed, tenant and application blobs now include their own names (`"tenant"` for tenants and `"application", "instance"` for applications (which isn't right, I know ...)), making the recursive @including a bit easier.

**Comment by jonmv** [comment]
Oh, there were more tests affected by this ... hang on!

**Comment by jonmv** [comment]
There.

**Comment by bratseth** [comment]
👍

**Pull request merged**